kotest-assertions/kotest-assertions-json/src/jvmMain/kotlin/io/kotest/assertions/json/JsonMatchers.kt

```diff
@@ -1,81 +1,208 @@
+            if (this == null) {
+                throw failure("bad quantity of key-value pairs in ${this.representation} - expected: $expected but the receiver is null")
+            }
 
-    val passed = try {
-      JsonPath.read<String>(value, path) != null
-    } catch (t: PathNotFoundException) {
-      false
+            if (this.isBlank() || this.firstOrNull { !it.isWhitespace() } != '{') {
+                throw failure("bad quantity of key-value pairs in ${this.representation} - expected: $expected but the receiver is not a JSON object: ${this.representation}")
+            }
+
+            val actual: Int = JsonPath.read(this, "length()")
+
+            if (actual != expected) {
+                throw failure("bad quantity of key-value pairs in ${this.representation} - expected: $expected but was: $actual")
+            }
+        }
+    }
+}
+
+@OptIn(ExperimentalContracts::class, ExperimentalStdlibApi::class)
+inline infix fun <reified T> Json?.shouldContainJsonKeyAndValueOfSpecificType(path: JsonKey): T {
+    contract {
+        returns() implies (this@shouldContainJsonKeyAndValueOfSpecificType != null)
     }
 
-    return MatcherResult(
-        passed,
-        "$sub should contain the path $path",
-        "$sub should not contain the path $path"
-    )
-  }
+    val result: Any? = try {
+        JsonPath.read(this, path)
+    } catch (thrown: Throwable) {
+        thrown
+    }
+
+    return when (result) {
+        is T -> result
+
+        is Throwable -> throw failure(
+            "${this.representation} should contain the path $path",
+            cause = result
+        )
+
+        else -> throw failure(
+            "${this.representation} should contain an element with type ${typeOf<T>()} with the path $path " +
+                    "but it contains ${mapper.writeValueAsString(result).representation}."
+        )
+    }
+}
+
+@OptIn(ExperimentalContracts::class)
+infix fun Json?.shouldContainOnlyJsonKey(path: JsonKey): Json {
+    contract {
+        returns() implies (this@shouldContainOnlyJsonKey != null)
+    }
+
+    this shouldContainExactly 1.jsonKeyValueEntries
+
+    return this shouldContainJsonKey path
 }
 
-fun <T> String.shouldContainJsonKeyValue(path: String, value: T) = this should containJsonKeyValue(path, value)
-fun <T> String.shouldNotContainJsonKeyValue(path: String, value: T) = this shouldNot containJsonKeyValue(path, value)
-fun <T> containJsonKeyValue(path: String, t: T) = object : Matcher<String> {
-  override fun test(value: String): MatcherResult {
-    val sub = if (value.length < 50) value.trim() else value.substring(0, 50).trim() + "..."
-    return MatcherResult(
-        JsonPath.read<T>(value, path) == t,
-        "$sub should contain the element $path = $t",
-        "$sub should not contain the element $path = $t"
-    )
-  }
+@OptIn(ExperimentalContracts::class)
+inline infix fun <reified T> Json?.shouldContainOnlyJsonKeyAndValueOfSpecificType(path: JsonKey): T {
```

Comment:
This function name is huge!
Can we find a way to make it shorter?

Reply:
I agree but can't make up any shorter suggestions...

Reply:
I think I can just omit introducing this function, but introduce the one for checking and converting type: `str?.shouldBeOfJsonType(jsonType): T`. It will be even more functional.